pkg/controller/plan/kubevirt.go

```diff
@@ -969,6 +937,130 @@ func (r *KubeVirt) DeleteHookJobs(vm *plan.VMStatus) (err error) {
+	if err != nil {
+		return
+	}
+	return
+}
+
+// Get the OpenstackVolumePopulator CustomResource List.
+func (r *KubeVirt) getOpenstackPopulatorCrList() (populatorCrList v1beta1.OpenstackVolumePopulatorList, err error) {
+	populatorCrList = v1beta1.OpenstackVolumePopulatorList{}
+	err = r.Client.List(
+		context.TODO(),
+		&populatorCrList,
+		&client.ListOptions{
+			Namespace:     r.Plan.Spec.TargetNamespace,
+			LabelSelector: labels.SelectorFromSet(map[string]string{kMigration: string(r.Plan.Status.Migration.ActiveSnapshot().Migration.UID)}),
+		})
+	if err != nil {
+		return
+	}
+	return
+}
+
+// Get the OvirtVolumePopulator CustomResource List.
+func (r *KubeVirt) getOvirtPopulatorCrList() (populatorCrList v1beta1.OvirtVolumePopulatorList, err error) {
+	populatorCrList = v1beta1.OvirtVolumePopulatorList{}
+	err = r.Client.List(
+		context.TODO(),
+		&populatorCrList,
+		&client.ListOptions{
+			Namespace:     r.Plan.Spec.TargetNamespace,
+			LabelSelector: labels.SelectorFromSet(map[string]string{kMigration: string(r.Plan.Status.Migration.ActiveSnapshot().Migration.UID)}),
+		})
+	if err != nil {
+		return
+	}
+	return
+}
+
+// Get the OvirtVolumePopulator CustomResource based on the PVC name.
+func (r *KubeVirt) getOvirtPopulatorCr(name string) (populatorCr v1beta1.OvirtVolumePopulator, err error) {
+	populatorCr = v1beta1.OvirtVolumePopulator{}
+	err = r.Client.Get(context.TODO(), client.ObjectKey{Namespace: r.Plan.Spec.TargetNamespace, Name: name}, &populatorCr)
+	if err != nil {
+		return
+	}
+	return
+}
+
+// Get the Openstack image from the inventory based on the PVC.
+func (r *KubeVirt) getOpenstackImage(pvc core.PersistentVolumeClaim) (image *openstack.Image, err error) {
+	image = &openstack.Image{}
+	err = r.Source.Inventory.Find(image, ref.Ref{ID: pvc.Name})
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	return
+}
+
+// Delete OpenstackVolumePopulator CustomResource list.
+func (r *KubeVirt) deleteOpenstackVolumePopulatorCrList(vm *plan.VMStatus) (err error) {
+	populatorCrList, err := r.getOpenstackPopulatorCrList()
+	if err != nil {
+		return err
+	}
+	for _, populatorCr := range populatorCrList.Items {
+		err = r.DeleteObject(&populatorCr, vm, "Deleted OpenstackPopulator CR.", "OpenstackVolumePopulator")
+		if err != nil {
+			return err
+		}
+	}
+	return
+}
+
+// Delete OvirtVolumePopulator CustomResource list.
+func (r *KubeVirt) deleteOvirtPopulatorCrList(vm *plan.VMStatus) (err error) {
+	populatorCrList, err := r.getOvirtPopulatorCrList()
+	if err != nil {
+		return err
+	}
+	for _, populatorCr := range populatorCrList.Items {
+		err = r.DeleteObject(&populatorCr, vm, "Deleted OvirtPopulator CR.", "OvirtVolumePopulator")
+		if err != nil {
+			return err
+		}
+	}
+	return
+}
+
+// Delete any populator pods that belong to a VM migration.
+func (r *KubeVirt) DeletePopulatorPods(vm *plan.VMStatus) (err error) {
```

Comment:
can we say something like, if the CR has the migration label then get the pods by the migration label?

Reply:
ah yeah, sorry, I didn't provide the reason for this suggestion - I meant that in order to identify whether the populator pods are set with the label, we can check whether the CR is set with the migration label. I'm worried about cleanup of populator pods that were not set with the migration label (i.e., populator pods that were created by 2.4)

Reply:
yeah, understood. i did think of it as well when doing the "labels" update on the controller start. we don't need it on the populator pods. because - until this version, the populator pod is immediately deleted after completion. if it fails it loops forever. when this version will come in, if the user will archive the plan it will clean the populator CR, only then the populator pod will stop restarting - in that case he still may have leftovers such as PVCs. 
if we wish to overcome it, i think we may need to update the populator pods labels as well like this PR is doing to the CRs.

Reply:
I don't see why not to set the populator pods with a label that points to the migration - again, it eases debugging when we know the context that an entity was created in

so you rely on the populator pods being cleared in 2.4? one argument to separate out the limit on the number of restarts in case of failure is to backport it to 2.4 - let's not assume they are cleared

Reply:
done